logging: fs backend: only delete old files if necessary

rerickson1 opened the pull request:
During initialization of the FS backed, the oldest
file is always deleted.
Fix this to prevent throwing away useful historic logs.
If there is space left in the newest file, append to it instead
of creating a new file and removing the oldest.

rerickson1 commented:
Looks like the current CI failure isn't related to the changes in this PR.

stephanosio commented:
Rebased to fix CI failure

stephanosio commented on subsys/logging/log_backend_fs.c:
```diff
@@ -263,6 +263,7 @@ static int allocate_new_file(struct fs_file_t *file)
 	int curr_file_num;
 	struct fs_dirent ent;
 	char fname[MAX_PATH_LEN];
+	int file_size;
```
this should be `off_t`.

de-nordic commented on subsys/logging/log_backend_fs.c:
```diff
@@ -339,14 +340,31 @@ static int allocate_new_file(struct fs_file_t *file)
+			goto out;
+		}
+		file_size = fs_tell(file);
+		if (file_size < CONFIG_LOG_BACKEND_FS_FILE_SIZE) {
```
This should be comparison against some selected threshold, because there is no point to use the file if you can only fit, for example, one byte in it, because it will be immediately closed and new file will be allocated.
Also this code skips the logic that cleans out old logs in case where there is no space left on device (lines 383-401), and you may have still space left within file but you will  not be able to write anything anyway, before remove something or decide to overwrite some other file.

rerickson1 replied:
Actually I don't think a threshold is needed.
This code only runs once when the backend is initialized, thus skipping the file cleanup at this time is ok because the code will have never let extra files be created during previous runs.

Also, see lines [154-177](https://github.com/zephyrproject-rtos/zephyr/pull/49700/files#diff-aef25f582f45618fe5f01cd4b464d21a31965570d559356fd99ea5ab2d1c11bfR154-R177).
`allocate_new_file` is called twice. Once for the initial init which will run this code to skip deleting the oldest file, then again if the newest file isn't large enough for the write. At that point it will clean up extra files if needed. This addresses your points.

Also [this test](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/subsys/logging/log_backend_fs/src/log_fs_test.c#L189) passed CI proving extra files are handled properly.

rerickson1 replied:
@de-nordic  Can you you re-review? The code already handles your concerns.

rerickson1 committed:
logging: fs backend: only delete old files if necessary

During initialization of the FS backed, the oldest
file is always deleted.
Fix this to prevent throwing away useful historic logs.
If there is space left in the newest file, append to it instead
of creating a new file and removing the oldest.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

de-nordic reviewed (approved):
Looks OK.
